### PR TITLE
Compact syntax for event lists

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
Currently, the window event list struct only has one (default) constructor, which leads to a bunch of null-pointers having to be set for most windows. It's not obvious which events they represent, which leaves these lists hard to maintain. Unfortunately, we cannot use named initialisers. Therefore, I propose a replacement set of definitions for the event lists using functions, giving them a syntax [similar to OpenLoco](https://github.com/OpenLoco/OpenLoco/blob/9b419912746c1b4930b1165e6dd4c149aeb48704/src/OpenLoco/Windows/Options.cpp#L633-L643).

I generated all of this using [a Python script](https://gist.github.com/AaronVanGeffen/110509418eca7bd22f6fbfb93c1406d3), which hopefully reduces the chance of errors. From brief testing on my end, all appears to work correctly. However, we should test this properly before merging.

### CODE REVIEW

**Comment by ZehMatt** [comment]
Since you already touch quite a bit of the code would be nice to have them in PascalCase/camelCase where appropriate.

**Comment by AaronVanGeffen** [comment]
I agree that would be nice, but I think that is really something for another PR. Preferably, we would rename the event functions into namespaces at the same time.

**Comment by ZehMatt** [comment]
We can probably avoid explicit initialization with something like this: https://godbolt.org/z/PqaKPf , what do you think about that?

**Comment by AaronVanGeffen** [comment]
I think it would be great if we can avoid explicitly calling an initialisation function. Your snippet looks like a great start in this respect. However, I am not sure I can adapt it directly. The problem I'm seeing is that the event handlers are stored as a pointer:
```cpp
struct rct_window
{
    rct_window_event_list* event_handlers;
```

Currently the event lists are passed as follows:
```
        window = window_create_auto_pos(windowWidth, 157, &window_guest_overview_events, WC_PEEP, WF_RESIZABLE);
```

I am wondering where the constructor (below) would be called here. Can we avoid making the assignment more explicit than this?
```cpp
    rct_window_event_list(auto fn)
    {
        fn(*this);
    }
```

File: src/openrct2-ui/windows/Cheats.cpp [hunk 1]
```diff
@@ -1332,4 +1265,4 @@ static void window_cheats_set_page(rct_window* w, int32_t page)
     w->widgets[WIDX_BACKGROUND].bottom = maxY - 1;
     w->widgets[WIDX_PAGE_BACKGROUND].bottom = maxY - 1;
     w->Invalidate();
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/Cheats.cpp:1268]: EOL

File: src/openrct2-ui/windows/ClearScenery.cpp [hunk 2]
```diff
@@ -253,4 +238,4 @@ static void window_clear_scenery_paint(rct_window* w, rct_drawpixelinfo* dpi)
         screenCoords.y = window_clear_scenery_widgets[WIDX_PREVIEW].bottom + w->windowPos.y + 5 + 27;
         gfx_draw_string_centred(dpi, STR_COST_AMOUNT, screenCoords, COLOUR_BLACK, &gClearSceneryCost);
     }
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/ClearScenery.cpp:241]: EOL

File: src/openrct2-ui/windows/CustomCurrency.cpp [hunk 3]
```diff
@@ -274,4 +257,4 @@ static void custom_currency_window_paint(rct_window* w, rct_drawpixelinfo* dpi)
                 + ScreenCoordsXY{ window_custom_currency_widgets[WIDX_AFFIX_DROPDOWN].left + 1,
                                   window_custom_currency_widgets[WIDX_AFFIX_DROPDOWN].top });
     }
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/CustomCurrency.cpp:260]: EOL

File: src/openrct2-ui/windows/DebugPaint.cpp [hunk 4]
```diff
@@ -184,4 +165,4 @@ static void window_debug_paint_invalidate(rct_window* w)
 static void window_debug_paint_paint(rct_window* w, rct_drawpixelinfo* dpi)
 {
     window_draw_widgets(w, dpi);
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/DebugPaint.cpp:168]: EOL

File: src/openrct2-ui/windows/DemolishRidePrompt.cpp [hunk 5]
```diff
@@ -240,4 +208,4 @@ static void window_ride_refurbish_paint(rct_window* w, rct_drawpixelinfo* dpi)
         ScreenCoordsXY stringCoords(w->windowPos.x + WW / 2, w->windowPos.y + (WH / 2) - 3);
         gfx_draw_string_centred_wrapped(dpi, gCommonFormatArgs, stringCoords, WW - 4, stringId, COLOUR_BLACK);
     }
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/DemolishRidePrompt.cpp:211]: EOL

File: src/openrct2-ui/windows/Dropdown.cpp [hunk 6]
```diff
@@ -471,4 +451,4 @@ void window_dropdown_show_colour(rct_window* w, rct_widget* widget, uint8_t drop
 uint32_t dropdown_get_appropriate_image_dropdown_items_per_row(uint32_t numItems)
 {
     return numItems < std::size(_appropriateImageDropdownItemsPerRow) ? _appropriateImageDropdownItemsPerRow[numItems] : 8;
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/Dropdown.cpp:454]: EOL

File: src/openrct2-ui/windows/EditorBottomToolbar.cpp [hunk 7]
```diff
@@ -522,4 +503,4 @@ void window_editor_bottom_toolbar_paint(rct_window* w, rct_drawpixelinfo* dpi)
             gfx_draw_string_centred(dpi, stringId, { textX, textY + 10 }, textColour, nullptr);
         }
     }
-}
+}
```
Comment (Broxzier) [inline src/openrct2-ui/windows/EditorBottomToolbar.cpp:506]: EOL

File: src/openrct2-ui/windows/EditorInventionsList.cpp [hunk 8]
```diff
@@ -846,4 +822,4 @@ static std::pair<rct_string_id, Formatter> window_editor_inventions_list_prepare
     return std::make_pair(drawString, ft);
 }
 
-#pragma endregion
+#pragma endregion
```
Comment (Broxzier) [inline src/openrct2-ui/windows/EditorInventionsList.cpp:825]: EOL

**Review by Broxzier — CHANGES_REQUESTED** [review]
Left a few comments on removed EOLs, this happened automatically by your editor, I presume. `insert_final_newline` is set in our `.editorconfig` file, so your editor seems to ignore this.

**Comment by AaronVanGeffen** [comment]
@Broxzier Ah, you're right, I hadn't noticed. This is entirely due to my Python script omitting them in its output — an oversight. I will amend. (It's a pity `clang-format` doesn't tackle this.)

**Comment by ZehMatt** [comment]
> I think it would be great if we can avoid explicitly calling an initialisation function. Your snippet looks like a great start in this respect. However, I am not sure I can adapt it directly. The problem I'm seeing is that the event handlers are stored as a pointer:
> 
> ```c++
> struct rct_window
> {
>     rct_window_event_list* event_handlers;
> ```
> 
> Currently the event lists are passed as follows:
> 
> ```
>         window = window_create_auto_pos(windowWidth, 157, &window_guest_overview_events, WC_PEEP, WF_RESIZABLE);
> ```
> 
> I am wondering where the constructor (below) would be called here. Can we avoid making the assignment more explicit than this?
> 
> ```c++
>     rct_window_event_list(auto fn)
>     {
>         fn(*this);
>     }
> ```

The construction should happen at the beginning of the application start, static initialization happens via a list the compiler constructed and iterates thru the constructors before any code is usually called, so that should be fine.

**Comment by AaronVanGeffen** [comment]
@ZehMatt Unfortunately, it seems your example only works if C++20 concepts are available. When I change to C++17 in Godbolt, it no longer compiles:

```
error: use of ‘auto’ in parameter declaration only available with ‘-fconcepts-ts’ [-Werror]
  245 |     rct_window_event_list(auto fn)
      |                           ^~~~
```

Any ideas?

**Comment by ZehMatt** [comment]
> @ZehMatt Unfortunately, it seems your example only works if C++20 concepts are available. When I change to C++17 in Godbolt, it no longer compiles:
> 
> ```
> error: use of ‘auto’ in parameter declaration only available with ‘-fconcepts-ts’ [-Werror]
>   245 |     rct_window_event_list(auto fn)
>       |                           ^~~~
> ```
> 
> Any ideas?

https://godbolt.org/z/Gv7Gn6 😏

**Comment by AaronVanGeffen** [comment]
I think I've got it. Even more impressive diffstat now :+1:

**Review by ZehMatt — APPROVED** [review]

**Pull request merged**